sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -113,31 +113,107 @@ object ScalaJSPluginInternal {
 
-        val outFile = AtomicWritableFileVirtualJSFile(output)
-        CacheUtils.cached(cp.version, outFile, Some(taskCache)) {
-          toolsIO.concatFiles(outFile, cp.jsLibs.map(getLib))
-        }
+        FileFunction.cached(s.cacheDirectory, FilesInfo.lastModified,
+            FilesInfo.exists) { _ => // We don't need the files
+
+          IO.createDirectory(output.getParentFile)
+
+          val outFile = AtomicWritableFileVirtualJSFile(output)
+          toolsIO.concatFiles(outFile, resolvedDeps.map(getLib))
+
+          Set(output)
+        } (realFiles.toSet)
 
         output
       }
     }
   }
 
-  private def scalaJSOptimizerSetting(key: TaskKey[_]): Setting[_] = (
-      scalaJSOptimizer in key := {
+  /** Settings for the production key (e.g. fastOptJS) of a given stage */
+  private def scalaJSStageSettings(stage: Stage, key: TaskKey[File]): Seq[Setting[_]] = Seq(
+      scalaJSLinker in key := {
+        val opts = (scalaJSOptimizerOptions in key).value
+
         val semantics = (scalaJSSemantics in key).value
         val outputMode = (scalaJSOutputMode in key).value
-        if ((scalaJSOptimizerOptions in key).value.parallel)
-          new ScalaJSOptimizer(semantics, outputMode, ParIncOptimizer.factory)
-        else
-          new ScalaJSOptimizer(semantics, outputMode, IncOptimizer.factory)
+        val withSourceMap = (emitSourceMaps in key).value
+
+        val relSourceMapBase = {
+          if ((relativeSourceMaps in key).value)
+            Some((artifactPath in key).value.getParentFile.toURI())
+          else
+            None
+        }
+
+        val frontendConfig = LinkerFrontend.Config()
+          .withBypassLinkingErrors(opts.bypassLinkingErrors)
+          .withCheckIR(opts.checkScalaJSIR)
+          .withDisableOptimizer(opts.disableOptimizer)
+
+        val backendConfig = LinkerBackend.Config()
+          .withRelativizeSourceMapBase(relSourceMapBase)
+          .withCustomOutputWrapper(scalaJSOutputWrapper.value)
+          .withPrettyPrint(opts.prettyPrintFullOptJS)
+
+        val newLinker = { () =>
+          Linker.linker(semantics, outputMode, withSourceMap, opts.parallel,
+              opts.useClosureCompiler, frontendConfig, backendConfig)
+        }
+
+        new ClearableLinker(newLinker, opts.batchMode)
+      },
+
+      scalaJSLinkingUnit in key := {
+        val s = (streams in key).value
+        val linker = (scalaJSLinker in key).value
+        val ir = scalaJSIR.value.data
+
+        linker.linkUnit(ir, s.log)
+      },
+
+      key := {
+        val s = (streams in key).value
+        val log = s.log
+        val irInfo = (scalaJSIR in key).value
+        val realFiles = irInfo.get(sourceFiles).get
+        val ir = irInfo.data
+        val output = (artifactPath in key).value
+
+        FileFunction.cached(s.cacheDirectory, FilesInfo.lastModified,
+            FilesInfo.exists) { _ => // We don't need the files
+
+          log.info(s"${stage.name} optimizing $output")
```

Comment:
This string interpolation is a big "dangerous". Depending on what `stage.name` is, this sentence might make no sense. Consider simply pattern matching over the possible `Stage`s (they're sealed).